### DISCUSSION TITLE
Godot should have a separate NRP render engine.

### DISCUSSION THREAD

**Discussion by ca3games** [godotengine/godot-proposals#4267] [3D]
Godot should have a separate NPR render engine, from their main PBR engine, because a NPR render engine has total diferent use case, performance considerations and goals than a PBR render engine.
so is not as simply as just using shaders with a PBR engine.

So far a NPR engine should have a solid toon shader, which could easily allow things like create a ramp color toon shader using a 2D texture, as to be able to easily draw in the texture the normal angle being used to compare the thresshold of the light angle to generate proper lighting that looks like artist hand made shadows over realistic more mechanical methods.
Also a NPR engine should have some automatic form to easily generate the normals of an anime head, since heads are an unique case that has to be handled with care, and manually tweaking the face normals in blender is a cumbersome solution, godot NPR render engine should have some setting where is similar to blender copy normals modifier, which uses a proxy simpler shape to project the normals of the face instead of using the mesh topology, maybe a simpler solution would be some user custom box or basic points to manually fit an already hand made face with good anime faces normal topology that works in most cases like GGXRd or other anime 3D fighting games being made by Arc system work.
This could totally save effort to less skilled amateur developers and help them get a correct solution to the anime faces issue in toon shading.

Also a NRP engine should have the option to add legacy materials like pong or other material solutions used in the mid 90s.
Another big feature is something similar to blender freestyle, which creates toon drawing lines based on setting paratemeters.
This could be used to create manga B&W background like photo inked renders.
Another feature could be a pixel art renderer for 3D meshes that store angle information in a voxel like data structure to tweak the final frame render of the mesh based on camera angle to manually reduce the artifacts that happens on 3D pixel art renders.

As you can see, NPR render engines are something beyond a basic toon shader with their own set of features needed and diferent design considerations.
Not as simple as just use a PBR engine with a toon shader.

**Comment by Calinou** [comment]
This is already being tracked in https://github.com/godotengine/godot-proposals/issues/484 and https://github.com/godotengine/godot-proposals/issues/1620. Note that it does *not* require replacing the entire rendering engine.

You may want to look at the existing toon shaders too: https://godotshaders.com/?s=toon
